### PR TITLE
Return the Host header value from `RequestHeaders.authority()` if `:a…

### LINKED ISSUE #2846: ServiceRequestContext.routingContext().hostname() yields different result depending on h2 vs http1.1
I'm having weird behaviour with ServiceRequestContext.routingContext().hostname(). When *passed through HAProxy* gives me
a) for http1.1 the hostname(i.e. what running `hostname` in a shell would give you)
b) for h2 what you put in the Host: header

`a` is never exhibited when curling Armeria directly without HAproxy. (using `curl --resolve yxvanus.staging.example.com:443:x.x.x.x`)

There is no difference when upgrading to 0.98.6 -> 0.99.7.

The actual discussion:
https://app.slack.com/client/T1NH00FK9/C1NGPBUH2/thread/C1NGPBUH2-1593150821.017200

curl https://yxvanus.staging.example.com
curl https://yxvanus.staging.example.com --http1.1

The actual requests as seen from the haproxy outside of it:
```
haproxy: 00001394:web.accept(0008)=000d from [66.249.93.192:54202] ALPN=<none>
haproxy: 00001394:web.clireq[000d:ffffffff]: GET /?cachebuster=1560797474 HTTP/1.1
haproxy: 00001394:web.clihdr[000d:ffffffff]: host: yxvanus.staging.example.com
haproxy: 00001394:web.clihdr[000d:ffffffff]: accept: */*
haproxy: 00001394:web.clihdr[000d:ffffffff]: user-agent: Mozilla/5.0 (compatible; Google-Site-Verification/1.0)
haproxy: 00001394:web.clihdr[000d:ffffffff]: accept-encoding: gzip,deflate,br
haproxy: 00001394:armeria.srvrep[000d:000f]: HTTP/2.0 303
haproxy: 00001394:armeria.srvhdr[000d:000f]: location: https://www.example.com
haproxy: 00001394:armeria.srvhdr[000d:000f]: content-length: 0
haproxy: 00001394:armeria.srvhdr[000d:000f]: strict-transport-security: max-age=31536000; includeSubDomains; preload
haproxy: 00001394:armeria.srvhdr[000d:000f]: x-rest-version: 0.0.1-YN-1-SNAPSHOT
haproxy: 00001394:armeria.srvhdr[000d:000f]: x-robots-tag: noindex
haproxy: 00001394:armeria.srvhdr[000d:000f]: server: Armeria/0.98.6
haproxy: 00001394:armeria.srvhdr[000d:000f]: date: Fri, 26 Jun 2020 01:41:20 GMT
haproxy: 00001394:armeria.srvcls[000d:000f]
haproxy: 00001394:armeria.clicls[000d:000f]
haproxy: 00001394:armeria.closed[000d:000f]
```

```
haproxy: 00003e0a:web.clireq[000e:ffffffff]: GET https://yxvanus.staging.example.com/ HTTP/2.0
haproxy: 00003e0a:web.clihdr[000e:ffffffff]: user-agent: curl/7.68.0
haproxy: 00003e0a:web.clihdr[000e:ffffffff]: accept: */*
haproxy: 00003e0a:web.clihdr[000e:ffffffff]: host: yxvanus.staging.example.com
haproxy: 00003e0c:web.closed[0012:ffffffff]
haproxy: 00003e0a:armeria.srvrep[000e:0010]: HTTP/2.0 200 
haproxy: 00003e0a:armeria.srvhdr[000e:0010]: content-type: text/html; charset=utf-8
haproxy: 00003e0a:armeria.srvhdr[000e:0010]: content-length: 5092
haproxy: 00003e0a:armeria.srvhdr[000e:0010]: strict-transport-security: max-age=31536000; includeSubDomains; preload
haproxy: 00003e0a:armeria.srvhdr[000e:0010]: x-rest-version: 0.0.1-YN-1-SNAPSHOT
haproxy: 00003e0a:armeria.srvhdr[000e:0010]: x-robots-tag: noindex
haproxy: 00003e0a:armeria.srvhdr[000e:0010]: server: Armeria/0.98.6
haproxy: 00003e0a:armeria.srvhdr[000e:0010]: date: Fri, 26 Jun 2020 05:55:57 GMT
haproxy: 00003e0a:armeria.srvcls[000e:0010]
haproxy: 00003e0a:armeria.clicls[000e:0010]
haproxy: 00003e0a:armeria.closed[000e:0010]
```

### PR DESCRIPTION
**Pull request by minwoox**
…uthority` is null

Motivation:
Currently, we translate `Host` header to `:authority` header when the protocol of a request is HTTP/1.1.
However, it's not expected behavior and a user still wants to get the `Host` header.
https://tools.ietf.org/html/rfc7540#section-8.1.3

So we should not translate header but just use the `Host` header value from `RequestHeaders.authority()`
when `:authority` header value is null in order to provide a unified getter for the authority.

Modifications:
- Return the `Host` header value from `RequestHeaders.authority()` when `:authority` header value is null.
- Add `Host` header if HTTP/1.1 headers does not have.

Result:
- Close #2846
- You can now get `Host` header from Armeria server if the request is made using HTTP/1.1

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java [hunk 1]
```diff
@@ -269,9 +271,12 @@ private void autoFillSchemeAndAuthority() {
...
         final String authority = endpoint != null ? endpoint.authority() : "UNKNOWN";
         if (headers.scheme() == null || !authority.equals(headers.authority())) {
-            unsafeUpdateRequest(req.withHeaders(headers.toBuilder()
-                                                       .authority(authority)
-                                                       .scheme(sessionProtocol())));
+            final RequestHeadersBuilder headersBuilder =
+                    headers.toBuilder().authority(authority).scheme(sessionProtocol());
+            if (headersBuilder.contains(HttpHeaderNames.HOST)) {
+                headersBuilder.remove(HttpHeaderNames.HOST);
+            }
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java:278]: Could just call `.remove()` without `.contains()`

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java]: Thanks fixed. 😉

File: core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttpUtil.java [hunk 2]
```diff
@@ -676,6 +672,18 @@ public static void toArmeria(io.netty.handler.codec.http.HttpHeaders inHeaders,
         }
     }
 
+    // Use Netty's validation logic once https://github.com/netty/netty/pull/10380 is merged.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttpUtil.java:675]: `TODO(minwoox)`?

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttpUtil.java]: Fixed. 😉

**Comment by trustin** [comment]
Thanks, @tobias- for reporting this issue. 🙇

**Comment by minwoox** [comment]
Thanks for reviewing!

**Pull request merged**